File: spark/temperature_data.py
from pyspark.sql.types import *
from pyspark.sql.functions import *
import os

class TemperatureData():

    def __init__(self, df):
        self.df = df
        self.temp_schema = StructType(
            [
                StructField("temperature", DoubleType(), True),
                StructField("humidity", DoubleType(), True),
                StructField("timestamp", DoubleType(), True)
            ]
        )

    def write_jdbc(self, df, epoch_id):
        # Transform and write batchDF
        df.persist()
        # df = df.withColumn(
        #     "id", monotonically_increasing_id()
        # )  # adding a db identifier
        df.write.format("jdbc").mode("append").options(
            url=os.getenv("CX_DB_URL"),
            dbtable="temperature",
            driver=os.getenv("CX_DB_DRIVER"),
            user=os.getenv("CX_DB_USER"),
            password=os.getenv("CX_DB_PWD"),
        ).save()
        df.unpersist()

    def start_working(self):
        sensor = self.df.filter("topic == 'temperature'")
        sensor = sensor.select(from_json(sensor.value, self.temp_schema).alias("data")).select(
            "data.*"
        )

        # converting timestamp column to Timestamp type, somehow the conversion doesn't work from the start.
        sensor = sensor.withColumn("timestamp", sensor["timestamp"].cast(TimestampType(
        )))

        ds = sensor.writeStream.foreachBatch(self.write_jdbc).start()
        return ds

File: aria-spark/src/main/java/io/codextension/aria/TemperatureData.java
package io.codextension.aria;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.functions;
import org.apache.spark.sql.streaming.StreamingQuery;
import org.apache.spark.sql.types.*;

import java.util.Map;

public class TemperatureData extends SensorData {

    private final StructType tempSchema;
    private Dataset<Row> ds;

    public TemperatureData(Dataset<Row> ds, Map<String, String> variables) {
        super(variables,"temperature");
        this.ds = ds;
        StructField[] fields = new StructField[3];
        fields[0] = new StructField("temperature", DataTypes.DoubleType, true, Metadata.empty());
        fields[1] = new StructField("humidity", DataTypes.DoubleType, true, Metadata.empty());
        fields[2] = new StructField("timestamp", DataTypes.DoubleType, true, Metadata.empty());
        this.tempSchema = new StructType(fields);
    }

    @Override
    public StreamingQuery startWorking() {
        Dataset<Row> sensor = this.ds.filter("topic == 'temperature'");
        sensor = sensor.select(functions.from_json(sensor.col("value"), this.tempSchema).alias("data")).select("data.*");
        sensor = sensor.withColumn("timestamp", sensor.col("timestamp").cast(DataTypes.TimestampType));

        return sensor.writeStream().foreachBatch(this::writeJDBC).start();
    }
}
